### ISSUE TITLE
vfs_import error when importing large files in to a <fs/> node

### ISSUE THREAD

**Issue by trimpim** [genodelabs/genode#3655]
I want to import the content of a tar file in to a `<fs/>` node. This fails with the following error:
```
[init -> vfs] Error: failed to create <import> VFS node
[init -> vfs] Error: 	overwrite="true"
```

Debugging the problem revealed that an exception is thrown at https://github.com/genodelabs/genode/blob/master/repos/os/src/lib/vfs/fs_file_system.h#L514 when the file is "too big".

**Comment by trimpim** [comment]
@nfeske I have a fix that solves the problem for me, but I'm not happy with it. I think there should be a better solution for this, but I can't see it at the moment.

**Comment by nfeske** [comment]
Thanks @trimpim for reporting. To me it seems that a proper fix would be a write loop implemented in the vfs-import plugin.

**Comment by trimpim** [comment]
@nfeske Is there a place where I can look how such a write loop is implemented? Then I would update my patch accordingly.

**Comment by nfeske** [comment]
You may find inspiration at https://github.com/genodelabs/genode/blob/master/repos/gems/src/app/text_area/new_file.h

Eventually, a utility like this will become part of `os/vfs.h` but before doing that, we need to gather more practical experience.

**Comment by trimpim** [comment]
@nfeske I'm currently trying to implement the write loop in `lib/vfs/fs_file_system.h`. I'm not sure, if this is the location you intended the write loop to go, as the current code in the `_write()` method uses the `packet_stream` interface to write the data.

To copy and modify the code so it compiled was not that difficult. But when I execute the test, `handle.fs().write()` never returns.

**Comment by nfeske** [comment]
Sorry for not being precise enough. The write loop belongs in the vfs-import plugin that is only active at initialization time. The fs-file-system must not implement a write loop but always immediately return to the caller.

**Comment by trimpim** [comment]
@nfeske no problem. I think this is ready for your review now.

**Comment by nfeske** [comment]
From glancing over the patch, it looks very nice. Only I wonder, shouldn't you better cancel the outer loop when a write error occurs?

**Comment by trimpim** [comment]
@nfeske True, I have updated the second commit.

**Comment by nfeske** [comment]
Thanks for reworking the patch. I have one more question: You removed the `Vfs_handle::Guard` and `Flush_guard` variables. Can you explain why? AFAIK, without these guards in place, we will leak file handles and miss the triggering of sync operations.

```
Vfs_handle::Guard guard(dst_handle);
...
Flush_guard flush(env, *dst_handle);
```

**Comment by trimpim** [comment]
@nfeske this was not intentional, I have fixed it and rebased to latest master.

**Comment by nfeske** [comment]
Alright. Thanks for the update. I'll merge your new version to staging after the next master update, presumably on Thursday.

**Comment by nfeske** [comment]
I merged the commit to staging along with a few spelling fixes of the commit message.

@trimpim even though I appreciate your contribution, I am not entirely happy about the process. The amount of guidance needed and the attention required from my side for reviewing the intermediate results is certainly more effort than I'm comfortable with. Regardless of whether you or someone at Genode Labs changes any code of the framework infrastructure, the burden of quality control is on Genode Labs' side. By accepting contributions, we are taking over this responsibility. But if I have to fix spelling mistakes, repeatedly pointing out our coding-style rules, or hinting at glaring omissions like the missing closing of the file descriptors, my confidence slips. Maybe we can reach a different mode of working?

I very much appreciate that you successfully find workable solutions for problems and limitations you encounter within Genode and share them with us. For us, it is important to know about these limitations and requirements, to keep them in the back of our heads, and come up with sustainable solutions over time. Such a sustainable solution often looks quite different from an immediate fix that you created for your problem. You may take this issue as a good illustration. Sometimes a sustainable solution takes *much* more effort than expected, e.g., when it calls for a partial software redesign. It should not be approached hastily.

Whenever someone creates a pull request featuring an intermediate fix, I feel the social pressure to mold the intermediate fix into a sustainable solution because this would be a prerequisite for merging it. But this urge overthrows the priorities of my other work. If the topic is neither on the road map nor otherwise critical for my current working topics, I suffer under a priority inversion problem. Glimpsing at #3712, I cannot help but see this pattern repeat.

Long story short, would you be fine with me taking your intermediate fixes as mere illustrations but deferring the actual solution as a topic for Genode Labs to work on? It goes without saying that I will still try to provide assistance. But it would be a relief to not enter the unspoken agreement to craft a solution for upstream in a timely manner together.

**Comment by trimpim** [comment]
@nfeske Sorry for causing that many interruptions on your side, this was certainly not my intention! 

I'll improve further contributions in this regard.

That said I'm definitely fine with your suggestion to take my intermediate fixes as an illustration for the problem while deferring the actual solution as a topic for Genode Labs!

**Comment by nfeske** [comment]
I'm grateful for your understanding.
